graph/pull_v2.go

```diff
@@ -232,60 +234,101 @@ func (p *v2Puller) pullV2Tag(tag, taggedName string) (bool, error) {
+	// downloads in parallel. We need a separate loop here *without* returns so
+	// that we don't launch some goroutines and then exit before they all
+	// terminate.
+	for i := len(manifest.FSLayers) - 1; i >= 0; i-- {
+		download := &downloads[i] // Note: we need a pointer to the download.
+		img := download.img
 
 		// Check if exists
 		if p.graph.Exists(img.ID) {
 			logrus.Debugf("Image already exists: %s", img.ID)
+			out.Write(p.sf.FormatProgress(stringid.TruncateID(img.ID), "Already exists", nil))
 			continue
 		}
 
 		out.Write(p.sf.FormatProgress(stringid.TruncateID(img.ID), "Pulling fs layer", nil))
 
-		downloads[i].err = make(chan error)
-		go p.download(&downloads[i])
+		download.err = make(chan error)
+		go p.download(download)
 	}
 
-	var tagUpdated bool
+	// NOTE: We cannot return *at all* within this loop. We must wait for every
+	// download goroutine to terminate before returning. When we reach the end
+	// of this loop, and pullErred is true, then you have my permission to die.
+	var tagUpdated, pullErred bool
 	for i := len(downloads) - 1; i >= 0; i-- {
 		d := &downloads[i]
-		if d.err != nil {
-			if err := <-d.err; err != nil {
-				return false, err
-			}
+
+		if d.err == nil {
+			// The error channel was never created so this download must have
+			// been skipped due to this layer already existing.
+			continue
 		}
-		if d.layer != nil {
-			// if tmpFile is empty assume download and extracted elsewhere
-			defer os.Remove(d.tmpFile.Name())
-			defer d.tmpFile.Close()
-			d.tmpFile.Seek(0, 0)
-			if d.tmpFile != nil {
-
-				reader := progressreader.New(progressreader.Config{
-					In:        d.tmpFile,
-					Out:       out,
-					Formatter: p.sf,
-					Size:      d.size,
-					NewLines:  false,
-					ID:        stringid.TruncateID(d.img.ID),
-					Action:    "Extracting",
-				})
-
-				err = p.graph.Register(d.img, reader)
-				if err != nil {
-					return false, err
-				}
-
-				if err := p.graph.SetDigest(d.img.ID, d.digest); err != nil {
-					return false, err
-				}
-
-				// FIXME: Pool release here for parallel tag pull (ensures any downloads block until fully extracted)
-			}
-			out.Write(p.sf.FormatProgress(stringid.TruncateID(d.img.ID), "Pull complete", nil))
-			tagUpdated = true
-		} else {
-			out.Write(p.sf.FormatProgress(stringid.TruncateID(d.img.ID), "Already exists", nil))
+
+		// This waits for the download to finish successfully or fail. We need
+		// to wait for all downloads even if previous ones failed so that we
+		// wait for all of the goroutines to die.
+		if err := <-d.err; err != nil {
+			pullErred = true
+			out.Write(p.sf.FormatProgress(stringid.TruncateID(d.img.ID), fmt.Sprintf("Download failed: %v", err), nil))
+			// This download failed. There's nothing we can do. Skip.
+			continue
 		}
+
+		// The download finished successfully. A tmpFile field is *always* set
+		// on successful downloads.
+		defer os.Remove(d.tmpFile.Name())
```

Comment:
Doesn't this mean that there will be no remove for the first download that fails? (since the continue is above the defers?)